tokio/src/runtime/global.rs

```diff
@@ -30,26 +30,9 @@ where
     T: Future + Send + 'static,
     T::Output: Send + 'static,
 {
-    EXECUTOR.with(|current_executor| match current_executor.get() {
-        #[cfg(feature = "rt-threaded")]
-        State::ThreadPool(thread_pool_ptr) => {
-            let thread_pool = unsafe { &*thread_pool_ptr };
-            thread_pool.spawn(future)
-        }
-        State::Basic(basic_scheduler_ptr) => {
-            let basic_scheduler = unsafe { &*basic_scheduler_ptr };
-
-            // Safety: The `BasicScheduler` value set the thread-local (same
-            // thread).
-            unsafe { basic_scheduler.spawn(future) }
-        }
-        State::Empty => {
-            // Explicit drop of `future` silences the warning that `future` is
-            // not used when neither rt-* feature flags are enabled.
-            drop(future);
-            panic!("must be called from the context of Tokio runtime configured with either `basic_scheduler` or `threaded_scheduler`");
-        }
-    })
+    let spawn_handle = crate::runtime::context::ThreadContext::spawn_handle()
+    .expect("must be called from the context of Tokio runtime configured with either `basic_scheduler` or `threaded_scheduler`");
+    spawn_handle.spawn(future)
```

Comment:
ah looks like this is a regression, this should call based on the enum

Reply:
Yea if I'm understanding the code correctly it will still be called based on the configured spawner but it's a roundabout way of doing it. I'm going to try and figure out a different solution which lifts
`static EXECUTOR: Cell<State> = Cell::new(State::Empty)`
into the `ThreadContext` so we can avoid having to bump both the `ThreadContext` refcount and the spawner refcount.

Reply:
Maybe the basic_scheduler should maintain its own thread-local (for now at least) that tracks the current executor. Then, in `basic_scheduler::Spawner` check the thread-local there to see if it matches the `Spawner(&self)`. If so, go via the "local spawn" path.

Reply:
Thanks @carllerche, I think we can take advantage of the existing
`static ACTIVE: Cell<*const SchedulerPriv> = Cell::new(ptr::null())` in `basic_scheduler.rs` for this.

Reply:
On second thought, it seems that this case may already be handled by virtue of `SchedulerPriv::schedule` checking to see if it is the `ACTIVE` scheduler.

Reply:
@gardnervickers if that is the case, then it seems fine. Even better than now (there may be an existing bug 👼

Reply:
Thanks! I ended up dropping the entire `global` module and enum in favor of just using the scheduler stored in `ThreadContext` for dispatch seeing as `SchedulerPriv::schedule` has it's own fast path for queuing tasks on spawn. I think this should be fine but I would appreciate a sanity check as I might be missing something.

@carllerche I'm not sure I follow, is there something here that indicates that there might be an existing bug?